### PR TITLE
BUILD-433: Run builds in user namespaces without seccomp if BUILD_PRIVILEGED=false

### PR DESCRIPTION
**Pull request by nalind**
In the build controller, we never actually called `setupContainersNodeStorage()`, so remove it, and stop setting the environment variable to force storage to use the overlay driver, which currently requires privileges, leaning on https://github.com/openshift/builder/pull/220 and https://github.com/openshift/builder/pull/291 to have the builder container figure that out for itself.

~Make the mode which we apply to secret and configuration map volumes in build pods something that the build controller can pass down to its helper functions, so that it can be adjusted based on the strategy.  Continue to default them to 0600.~

Set the seccomp profile for build pods to "unconfined", since we depend on `unshare()` and CRI-O is attempting to move away from allowing it to be used in its default seccomp profile.

When the Build object's environment includes "BUILD_PRIVILEGED=false" or "BUILD_PRIVILEGED=0":
* turn off the "privileged" flag for build pods
* set "io.openshift.builder=" and "io.kubernetes.cri-o.Devices=/dev/fuse:rwm" annotations
* set "io.kubernetes.cri-o.userns-mode=auto:size=65536"

### CODE REVIEW

**Comment by nalind** [comment]
/retitle Run builds with "privileged: false" if BUILD_PRIVILEGED=false

**Comment by nalind** [comment]
/retitle Run builds in user namespaces if BUILD_PRIVILEGED=false

**Comment by adambkaplan** [comment]
/retitle BUILD-433: Run builds in user namespaces if BUILD_PRIVILEGED=false

File: pkg/build/controller/strategy/custom.go [hunk 1]
```diff
@@ -88,7 +88,8 @@ func (bs *CustomBuildStrategy) CreateBuildPod(build *buildv1.Build, additionalCA
 		serviceAccount = buildutil.BuilderServiceAccountName
 	}
 
-	privileged := true
+	securityContext := securityContextForBuild(strategy.Env)
+	secretsMode := int32(0600) // todo: change to corev1.SecretVolumeSourceDefaultMode (0644)?
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/custom.go:92]: Do we want to do this for `Custom` strategy builds? Outside of OSBS this isn't really used (though this may be a vote in favor of enabling this capability).

File: pkg/build/controller/strategy/util.go [hunk 2]
```diff
@@ -113,20 +113,20 @@ func setupDockerSocket(pod *corev1.Pod) {
 
 // mountConfigMapVolume is a helper method responsible for actual mounting configMap
 // volumes into a pod.
-func mountConfigMapVolume(pod *corev1.Pod, container *corev1.Container, configMapName, mountPath, volumeSuffix string, volumeSource *corev1.VolumeSource) {
-	mountVolume(pod, container, configMapName, mountPath, volumeSuffix, policy.ConfigMap, volumeSource)
+func mountConfigMapVolume(pod *corev1.Pod, container *corev1.Container, configMapName, mountPath, volumeSuffix string, volumeSource *corev1.VolumeSource, configMapMode int32) {
+	mountVolume(pod, container, configMapName, mountPath, volumeSuffix, policy.ConfigMap, volumeSource, configMapMode)
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/util.go:117]: We can be opinionated here and mount the ConfigMap with the default volume mode. ConfigMap volumes must be read-only in the container.

File: pkg/build/controller/strategy/util.go [hunk 3]
```diff
@@ -113,20 +113,20 @@ func setupDockerSocket(pod *corev1.Pod) {
...
 }
 
 // mountSecretVolume is a helper method responsible for actual mounting secret
 // volumes into a pod.
-func mountSecretVolume(pod *corev1.Pod, container *corev1.Container, secretName, mountPath, volumeSuffix string, volumeSource *corev1.VolumeSource) {
-	mountVolume(pod, container, secretName, mountPath, volumeSuffix, policy.Secret, volumeSource)
+func mountSecretVolume(pod *corev1.Pod, container *corev1.Container, secretName, mountPath, volumeSuffix string, volumeSource *corev1.VolumeSource, secretMode int32) {
+	mountVolume(pod, container, secretName, mountPath, volumeSuffix, policy.Secret, volumeSource, secretMode)
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/util.go:123]: Ditto - we can be opinionated and mount with the default mode for Secrets

File: pkg/build/controller/strategy/util.go [hunk 4]
```diff
@@ -113,20 +113,20 @@ func setupDockerSocket(pod *corev1.Pod) {
...
 }
 
 // mountCSIVolume is a helper method responsible for actual mounting csi
 // volumes into a pod.
-func mountCSIVolume(pod *corev1.Pod, container *corev1.Container, volumeName, mountPath, volumeSuffix string, volumeSource *corev1.VolumeSource) {
-	mountVolume(pod, container, volumeName, mountPath, volumeSuffix, policy.CSI, volumeSource)
+func mountCSIVolume(pod *corev1.Pod, container *corev1.Container, volumeName, mountPath, volumeSuffix string, volumeSource *corev1.VolumeSource, csiVolumeMode int32) {
+	mountVolume(pod, container, volumeName, mountPath, volumeSuffix, policy.CSI, volumeSource, csiVolumeMode)
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/util.go:129]: CSI volumes do not have an associated mode, pass 0 here.

File: pkg/build/controller/strategy/util.go [hunk 5]
```diff
@@ -148,7 +148,6 @@ func mountVolume(pod *corev1.Pod, container *corev1.Container, objName, mountPat
 			break
 		}
 	}
-	mode := int32(0600)
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/util.go:151]: We should probably change this to the default of 0644 and leave it at that.

@bparees I found this commit set the mode to 0600 - do you recall why?
https://github.com/openshift/openshift-controller-manager/commit/b4cee01a40fdba76d3aa90dbb773da6ccf433258

**Review by adambkaplan — COMMENTED** [review]
My first pass - I am a bit concerned regarding the refactoring around volume modes - this should be addressed in a separate PR. Is it possible to revert these changes?

File: pkg/build/controller/strategy/util.go (hunk 5 above)
Comment (bparees) [reply pkg/build/controller/strategy/util.go:151]: i don't recall why we did it specifically, but logically it minimizes who can read the mounted secret and we know there's no reason for any additional `group` or `other` permissions to be granted.  Given these are secrets we're mounting, keeping them as secure as possible seems like the right permission setup.

(it doesn't matter *that* much since the only process running in the pod is our process which has access to the content, but hypothetically it removes some additional exposure avenues)

Comment (bparees) [reply pkg/build/controller/strategy/util.go:151]: i'll also note that it looks like this commit actually went in back in august 2017 (the github ui makes it look like it was 2019).  So it's been this way for a looooong time.

File: pkg/build/controller/strategy/util.go (hunk 4 above)
Comment (nalind) [reply pkg/build/controller/strategy/util.go]: Ah, good to know, thanks!

File: pkg/build/controller/strategy/util.go (hunk 5 above)
Comment (bparees) [reply pkg/build/controller/strategy/util.go]: full history on this can be seen here:
https://github.com/openshift/origin/pull/15248

tldr it wasn't a change from 644 to 0600, it was net new code that replaced existing code that used 0600.  The commit was part of a larger PR.

**Comment by nalind** [comment]
Removed the `error` that `setupBuilderAutonsUser()` could previously return that was always `nil`.

**Review by adambkaplan — APPROVED** [review]
/approve

**Comment by adambkaplan** [comment]
/cc @rolfedh 

For docs approved - draft https://github.com/openshift/openshift-docs/pull/44000

**Comment by adambkaplan** [comment]
/cc @jitendar-singh @prietyc123 

For QE approval - should be tested with https://github.com/openshift/builder/pull/291

**Comment by adambkaplan** [comment]
/cc @RickJWagner 

For PX approval

File: pkg/build/controller/strategy/util.go [hunk 6]
```diff
@@ -536,6 +518,50 @@ func addVolumeMountToContainers(conts []corev1.Container, mount corev1.VolumeMou
...
+	securityContext := &corev1.SecurityContext{
+		Privileged: &privileged,
+	}
+	if !privileged {
+		securityContext.Capabilities = &corev1.Capabilities{
+			Drop: []corev1.Capability{
+				"CAP_KILL",
+				"CAP_MKNOD",
+			},
+		}
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/util.go:544]: Do we need to run as uid/gid 0?
```suggestion
		securityContext.Capabilities = &corev1.Capabilities{
			Drop: []corev1.Capability{
				"CAP_KILL",
				"CAP_MKNOD",
			},
		}
		securityContext.RunAsUser = 0
		securityContext.RunAsGroup = 0
```

Comment (nalind) [reply pkg/build/controller/strategy/util.go:544]: That's the default for the struct, but you're right that there's no harm in making it explicit in the code.  Since that's what we're requesting whether or not we're privileged, should I just set these fields outside of the `if !privileged` block?

Comment (nalind) [reply pkg/build/controller/strategy/util.go:544]: Oh, those are pointers.  Got it.

**Review by RickJWagner — APPROVED** [review]
No suggested changes, beyond what's already under discussion.

**Comment by nalind** [comment]
> My 4.11 node only has entries for "core" in those files. That's confusing, because I had assumed that https://bugzilla.redhat.com/show_bug.cgi?id=2038827 was fixed for 4.11 when it was fixed in 4.10, and the "containers" entries are indeed there when I examine a 4.10.3 node running on CRC 2.0.1.

@haircommander's adding them to 4.11 in cri-o-1.24.0-9.rhaos4.11.git77718dc.el8.

**Review by rolfedh — APPROVED** [review]
/lgtm from a docs standpoint

**Comment by nalind** [comment]
* Have reverted to making `securityContext.RunAsUser` and `securityContext.RunAsGroup` implicit for unprivileged builds - if they're set explicitly, the service account token's permissions become 0600 instead of 0644, and the builder can't read them.
* Setting up a user namespace with the current set of UIDs/GIDs mapped to themselves (or more specifically, UID 0) to get `CAP_SYS_ADMIN` over their own new mount namespaces requires `CAP_SETFCAP` starting in upstream kernel 5.12 and RHEL kernel 4.18.0-312.el8.  Added it to the unprivileged build security context.

**Comment by nalind** [comment]
* When the pull secret is owned by root with mode `0o600`, an unprivileged builder can't read it.  Tweaked `pkg/build/controller/strategy.mountVolume()` to use mode `0o644` when the security context isn't privileged.

**Comment by adambkaplan** [comment]
@coreydaley if you'd like, I can set up this repo for "no-feature-freeze" so you don't need a BZ to merge.

**Comment by coreydaley** [comment]
@adambkaplan That would be great, thanks!
